### PR TITLE
Various optimizations for AI performance and not only

### PR DESCRIPTION
**Pull request by DjWarmonger**
- Reduced a number of slow map tiles iterations based on profiler results
- Relaxed some mutex locks so they are not held when printing logs

### CODE REVIEW

File: Global.h [hunk 1]
```diff
@@ -157,6 +157,7 @@ static_assert(sizeof(bool) == 1, "Bool needs to be 1 byte in size.");
 #include <boost/thread.hpp>
 #include <boost/variant.hpp>
 #include <boost/math/special_functions/round.hpp>
+#include <boost/multi_array.hpp>
```
Comment (alexvins) [inline Global.h]: GCC already can`t use our precompiled header - just crushes. We should cleanup Global.h ant not add more heavy stuff there.

File: AI/VCAI/VCAI.cpp [hunk 2]
```diff
@@ -2949,7 +2954,7 @@ void SectorMap::exploreNewSector(crint3 pos, int num, CCallback * cbp)
 {
 	Sector &s = infoOnSectors[num];
 	s.id = num;
-	s.water = cbp->getTile(pos)->isWater();
+	s.water = getTile(pos)->isWater();
```
Comment (alexvins) [inline AI/VCAI/VCAI.cpp]: getTile(pos)  may return nullptr

File: AI/VCAI/VCAI.cpp [hunk 3]
```diff
@@ -2960,7 +2965,7 @@ void SectorMap::exploreNewSector(crint3 pos, int num, CCallback * cbp)
 		ui8 &sec = retreiveTile(curPos);
 		if(sec == NOT_CHECKED)
 		{
-			const TerrainTile *t = cbp->getTile(curPos);
+			const TerrainTile *t = getTile(curPos);
 			if(!markIfBlocked(sec, curPos, t))
```
Comment (alexvins) [inline AI/VCAI/VCAI.cpp]: getTile(pos)  may return nullptr

File: lib/CGameInfoCallback.cpp [hunk 4]
```diff
@@ -455,6 +464,31 @@ const TerrainTile * CGameInfoCallback::getTile( int3 tile, bool verbose) const
...
+	boost::multi_array<TerrainTile*, 3> tileArray(boost::extents[width][height][levels]);
+	
+	for (size_t x = 0; x < width; x++)
+		for (size_t y = 0; y < height; y++)
+			for (size_t z = 0; z < levels; z++)
+			{
+				if (team->fogOfWarMap[x][y][z])
+					tileArray[x][y][z] = &gs->map->getTile(int3(x, y, z));
+				else
+					tileArray[x][y][z] = nullptr;
```
Comment (alexvins) [inline lib/CGameInfoCallback.cpp]: May be declare static TerrainTile constant for all invisible tiles, f.e. with BORDER terrain type.

File: AI/VCAI/VCAI.cpp (hunk 2 above)
Comment (DjWarmonger) [inline AI/VCAI/VCAI.cpp]: cbp->getTile(pos) could do as well, nothing has changed here.

Comment (alexvins) [inline AI/VCAI/VCAI.cpp]: True, but check is still desirable.

File: lib/CGameInfoCallback.h [hunk 5]
```diff
@@ -91,6 +91,7 @@ class DLL_LINKAGE CGameInfoCallback : public virtual CCallbackBase
 	const CMapHeader * getMapHeader()const;
 	int3 getMapSize() const; //returns size of map - z is 1 for one - level map and 2 for two level map
 	const TerrainTile * getTile(int3 tile, bool verbose = true) const;
+	shared_ptr<boost::multi_array<TerrainTile*, 3>> CGameInfoCallback::getAllVisibleTiles() const;
```
Comment (alexvins) [inline lib/CGameInfoCallback.h]: Shall be shared_ptr<boost::multi_array<TerrainTile*, 3>> getAllVisibleTiles() const;

**Comment by alexvins** [comment]
Closing, as there is newer PR with this features.

**Pull request closed without merging**